### PR TITLE
Sensor failure recovery, support calibration to the antenna, save onboard cals separately, cal on startup

### PR DESCRIPTION
**Pull request by dboulware**
This restores scos-sensor's ability to restart and recover from severe failures. In addition, support is added for a differential calibration file to allow for additional loss to move the reference point to the antenna. Onboard calibrations are modified to save to an onboard calibration file and the initial sensor calibration file is never altered to protect it from being corrupted during a shutdown or failure. Finally, settings are added to enable calibrations to optionally be performed on startup if the any calibrations occurred beyond a configurable amount of time in the past. Note: this will need requirements updating when scos-actions and scos-tekrsa PRs are merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -422,12 +422,11 @@ specific to the sensor you are using.
 
 ### Sensor Calibration File
 
-By default, scos-sensor will use `configs/default_calibration.json` as the sensor
-calibration file. However, if`configs/sensor_calibration.json` or
-`configs/sigan_calibration.json` exist they will be used instead of the default
-calibration file. Sensor calibration files allow scos-sensor to apply a gain based
-on a laboratory calibration of the sensor and may also contain other useful
-metadata that characterizes the sensor performance. For additional
+By default, scos-sensor will use `configs/default_sensor_calibration.json` as the sensor
```
Comment (jhazentia) [inline README.md:425]: It looks like this paragraph needs to be updated. The default file is now onboard_sensor_calibration.json. The sensor_calibration.json file will only be used if  onboard_sensor_calibration.json does not exist. Also should mention use and support of differential_calibration.json.

File: README.md [hunk 2]
```diff
@@ -422,12 +422,11 @@ specific to the sensor you are using.
...
+calibration file. However, if `configs/sensor_calibration.json` exists it will be used
+instead of the default calibration file. Sensor calibration files allow SCOS Sensor to
```
Comment (jhazentia) [inline README.md:427]: Suggest rewording this slightly to avoid confusion. Change `Sensor calibration files` to `The calibration files` to clarify that this sentence refers to calibration files in general and not specifically sensor_calibration.json.

File: env.template [hunk 3]
```diff
@@ -86,8 +87,12 @@ SSL_CA_PATH=scos_test_ca.crt
 SSL_CERT_PATH=sensor01.pem
 SSL_KEY_PATH=sensor01.pem
 
-USB_DEVICE=Tektronix
-
+# Calibration action selection
+#    The action specified here will be used to attempt an onboard
+#    sensor calibration on startup, if no onboard calibration data
+#    is available on startup. The specified action must be available.
+CALIBRATE_ON_STARTUP=true
```
Comment (jhazentia) [inline env.template:94]: Suggest removing CALIBRATE_ON_STARTUP as it is not used or modify the code to use it if there is a need to control this behavior.

File: env.template [hunk 4]
```diff
@@ -75,8 +72,12 @@ SCOS_SENSOR_GIT_TAG="$(git describe --tags)"
 # SECURITY WARNING: generate unique key with `manage.py generate_secret_key`
 SECRET_KEY="$(python3 -c 'import secrets; print(secrets.token_urlsafe(64))')"
 
+# Signal analyzer selection/setup
 SIGAN_CLASS=TekRSASigan
 SIGAN_MODULE=scos_tekrsa.hardware.tekrsa_sigan
+USB_DEVICE=Tektronix
+DEVICE_MODEL=RSA507A
+BASE_IMAGE=ghcr.io/ntia/scos-tekrsa/tekrsa_usb:0.2.3
```
Comment (jhazentia) [inline env.template:80]: Consider changing the 0.2.3 to `latest` to match scos-tekrsa README, but perhaps there is a need to keep it pinned to specific version?

**Review by jhazentia — CHANGES_REQUESTED** [review]
Looks good, just some minor comments

File: README.md (hunk 1 above)
Comment (aromanielloNTIA) [reply README.md]: Updated README in 5680801a6837a66ee28103b8b509b4eb42ac2602

File: README.md (hunk 2 above)
Comment (aromanielloNTIA) [reply README.md]: Updated README in 5680801a6837a66ee28103b8b509b4eb42ac2602

File: env.template (hunk 3 above)
Comment (aromanielloNTIA) [reply env.template]: Removed in fc8335b77ac1f6d1a9ba480ef33dba3642a75f40

File: README.md [hunk 5]
```diff
@@ -518,7 +525,7 @@ per second at several frequencies with a signal analyzer reference level setting
 
 When an action is run with the above calibration, SCOS will expect the action to have
 a sample_rate, frequency, and reference_level specified in the action config. The values
```
Comment (jhazentia) [inline README.md:527]: This example file also requires preamp_enable and attenuation

File: README.md [hunk 6]
```diff
@@ -420,47 +428,46 @@ specific to the sensor you are using.
...
+the calibration data. In the case of onboard calibration being generated from scratch,
+the startup calibration action's signal analyzer settings are used as the `calibration_parameters`.
+The calibration data is found directly within the `calibration_data` element and by
+default SCOS Sensor will not apply any additional gain. Typically, a sensor would be
+calibrated at particular sensing parameters. The calibration data for specific parameters
+should be listed within the `calibration_data` object and accessed by the values of the
+settings listed in the calibration_parameters element. For example, the sensor
+calibration below provides an example of a sensor calibrated at a sample rate of
+140000000 samples per second at several frequencies with a signal analyzer reference
+level setting of -25.
```
Comment (jhazentia) [inline README.md:470]: Consider mentioning preamp_enable and attenuation settings as these are also in the example below

File: README.md (hunk 5 above)
Comment (aromanielloNTIA) [reply README.md:513]: To resolve this, I opted to adjust the example to remove preamp_enable and attenuation as cal parameters. This makes a smaller example that is a bit easier to parse. (in 2eb839f4eec5d249aa2661b16b5e5a4103382b2f)

File: README.md (hunk 6 above)
Comment (aromanielloNTIA) [reply README.md:470]: See my other comment above- addressed this in 2eb839f4eec5d249aa2661b16b5e5a4103382b2f by modifying the example

File: src/initialization/sensor_loader.py [hunk 7]
```diff
@@ -55,45 +55,38 @@ def load_sensor(sensor_capabilities: dict) -> Sensor:
...
             logger.info("Running migrations. Not loading signal analyzer.")
-    except Exception as ex:
+    except BaseException as ex:
         logger.warning(f"unable to create signal analyzer: {ex}")
+        set_container_unhealthy()
 
+    # Create sensor before handling calibrations
     sensor = Sensor(
         signal_analyzer=sigan,
+        # TODO GPS Not Implemented
```
Comment (aromanielloNTIA) [inline src/initialization/sensor_loader.py:83]: Should this be addressed now? SCOS Sensor in its current state already does ever pass a GPS to the sensor constructor. We may want to create a new issue to address this in the future

Comment (aromanielloNTIA) [reply src/initialization/sensor_loader.py:83]: @jhazentia

Comment (jhazentia) [reply src/initialization/sensor_loader.py:83]: Working on this in #272

Comment (aromanielloNTIA) [reply src/initialization/sensor_loader.py:83]: Sounds good, we will leave it alone for now.

**Review by jhazentia — APPROVED** [review]
Looks good!
